src/Formatter/Formatter.php

```diff
@@ -227,4 +250,69 @@ public function getJs()
     {
         return $this->getComponent('js');
     }
+
+    protected function ensureRendererExists()
+    {
+        if (! static::$formatter) {
+            return;
+        }
+
+        $revision = $this->getRevision();
+
+        if (file_exists($this->cacheDir."/Renderer_$revision.php")) {
+            return;
+        }
+
+        $renderer = Arr::get(static::$formatter, 'renderer');
+
+        if (! empty($renderer)) {
+            file_put_contents($this->cacheDir."/Renderer_$revision.php", $renderer);
+        } else {
+            $this->finalize();
+        }
+
+        // Reload because finalizing might have generated a new one.
+        $renderer = Arr::get(static::$formatter, 'renderer');
+
+        if ($renderer && static::$formatter['renderer'] instanceof __PHP_Incomplete_Class) {
+            // Autoload the file from disk using a simple include, while suppressing errors.
+            @include $this->cacheDir."/Renderer_$revision.php";
+
+            // Reload the formatter again from cache to resolve the __PHP_Incomplete_Class
+            static::$formatter = $this->cache->get('flarum.formatter');
+        }
+    }
+
+    protected function finalize()
+    {
+        $formatter = $this->getConfigurator()->finalize();
+
+        preg_match('~^Renderer\_(?<revision>[^\.]+)$~', get_class($formatter['renderer']), $m);
```

Comment:
Can we make this regex a private method? It's repeated several times now. IMO it should just return the revision so we don't have to carry around the `$m` object